src/java.base/share/classes/java/math/BigInteger.java

```diff
@@ -3864,29 +3853,11 @@ public int getLowestSetBit() {
      *         representation of this BigInteger, <em>excluding</em> a sign bit.
      */
     public int bitLength() {
-        int n = bitLengthPlusOne - 1;
-        if (n == -1) { // bitLength not initialized yet
-            int[] m = mag;
-            int len = m.length;
-            if (len == 0) {
-                n = 0; // offset by one to initialize
-            }  else {
-                // Calculate the bit length of the magnitude
-                int magBitLength = ((len - 1) << 5) + bitLengthForInt(mag[0]);
-                 if (signum < 0) {
-                     // Check if magnitude is a power of two
-                     boolean pow2 = (Integer.bitCount(mag[0]) == 1);
-                     for (int i=1; i< len && pow2; i++)
-                         pow2 = (mag[i] == 0);
-
-                     n = (pow2 ? magBitLength - 1 : magBitLength);
-                 } else {
-                     n = magBitLength;
-                 }
-            }
-            bitLengthPlusOne = n + 1;
-        }
-        return n;
+        return signum < 0
+                // Check if magnitude is a power of two
+                && Integer.lowestOneBit(mag[0]) == mag[0]
+                && numberOfTrailingZeroInts() == mag.length - 1
+            ? magBitLength() - 1 : magBitLength();
```

Comment:
```suggestion
        int[] mag = this.mag;
        return magBitLength() 
                - (signum < 0
                   // Check if magnitude is a power of two
                   && Integer.lowestOneBit(mag[0]) == mag[0]
                   && numberOfTrailingZeroInts() == mag.length - 1
                   ? 1 : 0);
```

In this way, codeSize will drop from 52 to 45

Reply:
I think the suggestion is fine.

Reply:
Yes, but perhaps in the current version it's more clear what is the returned value according to the condition.

Reply:
Moreover, the current version avoid a subtraction in almost all cases, although probably it is negligible.